JP-2310: Move dark current to STCAL

kmacdonald-stsci opened the pull request:
This moves the dark current subtraction step to STCAL.  With a change in paths, the JWST unit tests pass.

dmggh commented on tests/test_dark_current.py:
```diff
@@ -0,0 +1,433 @@
+"""
```
Some of these tests are MIRI-specific; should they or the MIRI-specific functions they call be in the JWST repo instead of here in STCAL ? How strictly do we want to adhere to having no mission-specific code in STCAL ?

dmggh reviewed (commented):
Looks good. Minor comments, and a general question about mission-specific tests.

nden commented on CHANGES.rst:
```diff
@@ -1,3 +1,11 @@
+dark_current
+------------
+
+- Moved dark current code from JWST to STCAL.
```
```suggestion
- Moved dark current code from JWST to STCAL. [#63]
```

kmacdonald-stsci replied:
I made this change.

nden commented on src/stcal/dark_current/dark_class.py:
```diff
@@ -0,0 +1,87 @@
+
+class DarkData:
+    """
+    Class removing data model dependencies.
```
I think we need a one line sentence summarizing what the class represents, not why it was created.

kmacdonald-stsci replied:
I changed this comment.

nden commented on src/stcal/dark_current/dark_class.py:
```diff
@@ -0,0 +1,87 @@
+        """
+        Creates a class to remove data model dependencies in the internals of
+        the dark current code.  The data in this class are for the science
+        exposure data on which to do the dark current step.
```
Similar comment here - write a one sentence description of the class. If you need to say more add the longer description as a small paragraph after that.

nden commented on src/stcal/dark_current/dark_class.py:
```diff
@@ -0,0 +1,87 @@
+        the dark current code.  The data in this class are for the science
+        exposure data on which to do the dark current step.
+
+        Paramters
```
"Paramters" - "Parameters"

nden commented on src/stcal/dark_current/dark_class.py:
```diff
@@ -0,0 +1,87 @@
+
+        Paramters
+        ---------
+        science_model: data model
```
Here and everywhere else the `Parameters` section needs to be corrected.

Here's the style guide:
https://numpydoc.readthedocs.io/en/latest/format.html

Note, that there should be one space between the parameter name and `:`, followed by a space and a `data type`.
So the above line should be:

```
science_model : `ScienceData`
```

Note also that single backticks create a link in the rendered documentation, while double backticks don't. Each is useful at different times.

nden commented:
I have the same concern as @dmggh - this code does not provide a generic algorithm that will work with any data. Instead the input data is repackaged but the checks for MIRI, e.g. still exist.
 
As it stands now, there's code repetition - the treatment of groups within each integration. Could that be a function?

Please follow the docstrings style guide

https://numpydoc.readthedocs.io/en/latest/format.html

kmacdonald-stsci commented on src/stcal/dark_current/dark_sub.py:
```diff
@@ -0,0 +1,391 @@
+    """
+    Averages the individual frames of data in a dark reference
+    file to match the group structure of a science data set.
+    MIRI needs a seperate routine because the darks are integration dependent.
```
I corrected this spelling.

kmacdonald-stsci commented on src/stcal/dark_current/dark_sub.py:
```diff
@@ -0,0 +1,391 @@
+    # average for num_ints integrations (if science data only has
+    # 1 integration then there is no need to average the second integration
+    # of the dark refefence file)
+    num_ints = dint
```
I made the suggested changes to this comment.

kmacdonald-stsci commented on src/stcal/dark_current/dark_class.py:
```diff
@@ -0,0 +1,87 @@
+        ---------
+        dims: tuple
+            A tuple of integers to describe the dimensions of the arrays used
+            during the dark current step.
```
I expanded the comments to clarify its usage.

hbushouse commented on src/stcal/dark_current/dark_sub.py:
```diff
@@ -0,0 +1,391 @@
+    sci_nframes = science_data.exp_nframes
+    sci_groupgap = science_data.exp_groupgap
+
+    if instrument == 'MIRI':
```
It might be possible to remove the instrument-dependent code here by looking at len(dark_data.data.shape). If it's 3, then it's not MIRI data and drk_nints=1, drk_ngroups=shape[0]. If it's 4, then drk_nints=shape[0], drk_ngroups=shape[1].

hbushouse commented on src/stcal/dark_current/dark_sub.py:
```diff
@@ -0,0 +1,391 @@
+        # we average them with a seperate routine.
+
+        if instrument == 'MIRI':
+            averaged_dark = average_MIRIdark_frames(
```
I wonder if it's possible to consolidate the `average_MIRIdark_frames` and `average_dark_frames` into a single function, which just checks the dimensions of the dark_data (i.e. if it's 4-D, then handle multiple integrations, otherwise ignore the integrations axis), rather than having hardwired MIRI-specific functions and code hooks.

kmacdonald-stsci commented:
I removed references to the MIRI instrument and changed conditionals and function names to be based on data characteristics.  All tests, including JWST unit and regression tests pass.
